small/matras.c

```diff
@@ -485,3 +514,76 @@ matras_touch(struct matras *m, matras_id_t id)
 
 	return matras_touch_no_check(m, id);
 }
+
+/**
+ * Create the given matras allocator.
+ */
+void
+matras_allocator_create(struct matras_allocator *allocator,
+			size_t extent_size,
+			matras_alloc_func alloc_func,
+			matras_free_func free_func)
+{
+	allocator->extent_size = extent_size;
+	allocator->alloc_func = alloc_func;
+	allocator->free_func = free_func;
+	allocator->reserved_extents = NULL;
+	allocator->num_reserved_extents = 0;
+}
+
+/**
+ * Destroy the given matras allocator.
+ */
+void
+matras_allocator_destroy(struct matras_allocator *allocator)
+{
+	while (allocator->num_reserved_extents > 0) {
+		void *ext = allocator->reserved_extents;
+		allocator->reserved_extents = *(void **)ext;
+		allocator->free_func(allocator, ext);
+		allocator->num_reserved_extents--;
+	}
+}
+
+/**
+ * Allocate a new extent or take one from the reserved list.
+ */
+void *
+matras_allocator_alloc(struct matras_allocator *allocator)
+{
+	if (allocator->num_reserved_extents > 0) {
+		void *ext = allocator->reserved_extents;
+		allocator->reserved_extents = *(void **)ext;
+		allocator->num_reserved_extents--;
+		return ext;
+	}
+	return allocator->alloc_func(allocator);
+}
+
+/**
+ * Return the given extent to the external allocator.
+ */
+void
+matras_allocator_free(struct matras_allocator *allocator, void *ext)
+{
+	allocator->free_func(allocator, ext);
+}
+
+/**
+ * Request and put into the reserved list new extents until its size is @p
+ * count or greater. Returns 0 on success, -1 on memory error.
+ */
+int
+matras_allocator_reserve(struct matras_allocator *allocator, int count)
+{
+	assert(count > 0);
```

Comment:
I guess `>=` is better as `0` argument should be ok.

Reply:
Fixed.